File: ts/src/memory/MemoryRepository.ts
import { NotFoundError } from "../RepositoryError";
import {
  Model,
  Repository,
  ListResult,
  Page,
  Sort,
  Filter,
  StringId,
} from "../Repository";

import { DataToObjectId, Uuid } from "./Uuid";

export class MemoryRepository<TModel extends Model, TFilter extends Filter>
  implements Repository<TModel, TFilter>
{
  collection: Record<string, StringId<TModel>> = {};

  // TODO: Should we copy the object?
  _createId(document: TModel): StringId<TModel> {
    // @ts-expect-error TODO: Fix this type
    if (!document._id) {
      // @ts-expect-error TODO: Fix this type
      document._id = Uuid();
    }
    return document as any;
  }

  _insert<TData extends TModel[]>(...data: TData): DataToObjectId<TData> {
    const records = data.map(this._createId);
    this.collection = records.reduce(
      (collection, document) => ({
        ...collection,
        [document._id?.toString() || ""]: document,
      }),
      this.collection
    );
    return records.map(({ _id }) => _id) as any;
  }

  _clear(...data: TModel[]): string[] {
    this.collection = {};
    if (data.length) {
      return this._insert(...data);
    } else {
      return [];
    }
  }

  _getIds() {
    return Object.values(this.collection).map(({ _id }) => _id);
  }

  async getList(
    filter: TFilter,
    { offset = 0, limit = 0 }: Page = { offset: undefined, limit: undefined },
    // TODO: Fix this
    sort?: Sort<TModel>
  ): Promise<ListResult<StringId<TModel>>> {
    const data = Object.values(this.collection)
      .filter(
        (document) =>
          !Object.entries(filter)
            .filter(([_key, value]) => value !== undefined)
            .find(
              ([key, value]) => document[key as keyof typeof document] !== value
            )
      )
      .slice(offset, offset + limit);
    const count = data.length;
    return {
      count,
      data,
    };
  }

  async getById(_id: string): Promise<StringId<TModel> | null> {
    return this.collection[_id] ?? null;
  }

  async create(data: TModel): Promise<string> {
    const _id = Uuid();
    const record = {
      ...data,
      _id,
    };
    this.collection[_id] = record as any;
    return _id;
  }

  async update(_id: string, data: TModel): Promise<boolean> {
    const record = this.collection[_id];
    if (!record) {
      throw new NotFoundError();
    }
    const updatedRecord = {
      _id,
      ...data,
    };
    this.collection[_id.toString()] = updatedRecord;
    return true;
  }

  async delete(_id: string): Promise<void> {
    const record = this.collection[_id.toString()];
    if (!record) {
      throw new NotFoundError();
    }
    delete this.collection[_id.toString()];
  }
}

File: go-memory/memory_repository.go
package memory

import (
	"context"
	"sort"

	repository "github.com/cardboardrobots/repository/go"
)

type MemoryRepository[T any, Q repository.Query] struct {
	Items             map[string]T
	preInsertCallback func(value T, id string)
	compare           func(a T, b T, sort ...repository.Sort) bool
}

func NewMemoryRepository[T any, Q repository.Query](preInsertCallback func(value T, id string), compare func(a T, b T, sort ...repository.Sort) bool) *MemoryRepository[T, Q] {
	return &MemoryRepository[T, Q]{
		Items:             map[string]T{},
		preInsertCallback: preInsertCallback,
		compare:           compare,
	}
}

var _ repository.Repository[repository.Valid, repository.Query] = &MemoryRepository[repository.Valid, repository.Query]{}

func (c *MemoryRepository[T, Q]) GetList(ctx context.Context, query Q, page repository.Page, _sort ...repository.Sort) (repository.ListResult[T], error) {
	limit := page.Limit
	if limit == 0 {
		limit = 10
	}
	count, _ := c.Count(ctx)
	data := c.Slice()
	sort.Slice(data, func(i, j int) bool {
		a := data[i]
		b := data[j]
		return c.compare(a, b, _sort...)
	})
	data = data[page.Offset : limit-page.Offset]
	result := repository.NewListResult(count, data)
	return result, nil
}

func (c *MemoryRepository[T, Q]) Count(ctx context.Context) (int, error) {
	return len(c.Items), nil
}

func (c *MemoryRepository[T, Q]) Slice() []T {
	data := make([]T, len(c.Items))
	i := 0
	for _, value := range c.Items {
		data[i] = value
		i++
	}
	return data
}

func (c *MemoryRepository[T, Q]) GetById(ctx context.Context, id string) (T, error) {
	value, ok := c.Items[id]
	if !ok {
		var zero T
		return zero, repository.NewErrNotFound()
	}

	return value, nil
}

func (c *MemoryRepository[T, Q]) Create(ctx context.Context, value T) (string, error) {
	id := Uuid()
	c.preInsertCallback(value, id)
	c.Items[id] = value
	return id, nil
}

func (c *MemoryRepository[T, Q]) Replace(ctx context.Context, id string, value T) error {
	c.Items[id] = value
	return nil
}

func (c *MemoryRepository[T, Q]) Update(ctx context.Context, id string, value T) (bool, error) {
	_, ok := c.Items[id]
	if !ok {
		return false, repository.NewErrNotFound()
	}

	c.Items[id] = value
	return true, nil
}

func (c *MemoryRepository[T, Q]) Delete(ctx context.Context, id string) (bool, error) {
	_, ok := c.Items[id]
	if !ok {
		return false, repository.NewErrNotFound()
	}

	delete(c.Items, id)
	return true, nil
}
